File: easy/Trim_a_Binary_Search_Tree.py
'''
669.
Trim a Binary Search Tree
https://leetcode.com/problems/trim-a-binary-search-tree/
'''

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def trimBST(self, root: TreeNode, low: int, high: int) -> TreeNode:
        if not root:
            return None
        
        if root.val < low:          # 노드 값이 low보다 작으면 보다 값이 큰 우측 노드로 재귀
            return self.trimBST(root.right, low, high)
        
        if root.val > high:         # 노드 값이 high보다 크면 보다 값이 작은 좌측 노드로 재귀
            return self.trimBST(root.left, low, high)
        
        root.left = self.trimBST(root.left, low, high)      # 좌, 우측 서브트리에도 trimBST를 적용
        root.right = self.trimBST(root.right, low, high)
        
        return root

'''
Runtime: 40 ms, faster than 97.38% of Python3 online submissions for Trim a Binary Search Tree.
Memory Usage: 18.3 MB, less than 30.13% of Python3 online submissions for Trim a Binary Search Tree.
'''

File: medium/Trim_a_Binary_Search_Tree.java
/*
669.
Trim a Binary Search Tree
https://leetcode.com/problems/trim-a-binary-search-tree/
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode trimBST(TreeNode root, int low, int high) {
        if (root == null) {     // null 노드 처리
            return null;
        }
        
        if (root.val < low) {   // 현재 노드의 값이 low값보다 작으면 자신과 왼쪽 자식노드를 버리고 오른쪽 자식 노드로 순회
            return trimBST(root.right, low, high);
        }
        
        if (root.val > high) {      // 현재 노드의 값이 high보다 크면 자신과 오른쪽 자식노드를 버리고 왼쪽 자식 노드로 순회
            return trimBST(root.left, low, high);
        }
        
        root.left = trimBST(root.left, low, high);      // 현재 노드의 값이 범위 안에 들면 양쪽 자식 노드로 순회하여 자식 노드 관계를 재설정
        root.right = trimBST(root.right, low, high);
        return root;
    }
}

/*
Runtime: 0 ms, faster than 100.00% of Java online submissions for Trim a Binary Search Tree.
Memory Usage: 41.9 MB, less than 84.81% of Java online submissions for Trim a Binary Search Tree.
*/
